torch/_prims/rng_prims.py

```diff
@@ -132,5 +148,163 @@ def _philox_rand(
+
+    @run_and_save_rng_state.py_impl(DispatchKey.Autograd)
+    def impl_autograd(op, *args, **kwargs):
+        with torch._C._AutoDispatchBelowAutograd():
+            return run_and_save_rng_state(op, *args, **kwargs)
+
+    @run_and_save_rng_state.py_impl(DispatchKey.CUDA)
+    def impl_cuda(op, *args, **kwargs):
+        return torch.cuda.get_rng_state(), op(*args, **kwargs)
+
+    @run_and_save_rng_state.py_impl(DispatchKey.CPU)
+    def impl_cpu(op, *args, **kwargs):
+        return torch.get_rng_state(), op(*args, **kwargs)
+
+    @run_and_save_rng_state.py_impl(DispatchKey.BackendSelect)
+    def impl_backend_select(op, *args, **kwargs):
+        # with _ExcludeDispatchKeyGuard(DispatchKeySet(DispatchKey.BackendSelect)):
+        if kwargs.get("device"):
+            device = kwargs.get("device")
+            device_type = device if isinstance(device, str) else device.type  # type: ignore[union-attr]
+            if "cuda" in device_type:
+                impl = impl_cuda
+            elif "cpu" in device_type:
+                impl = impl_cpu
+            else:
+                raise NotImplementedError(
+                    f"Backend not available for device {device_type}"
+                )
+        else:
+            devices = {arg.device.type for arg in args if isinstance(arg, torch.Tensor)}
+            if any("cuda" in dev for dev in devices):
+                impl = impl_cuda
+            elif any("cpu" in dev for dev in devices):
+                impl = impl_cpu
+            else:
+                raise NotImplementedError(
+                    f"Backend not available for device {device_type}"
+                )
+        return impl(op, *args, **kwargs)
+
+    @run_and_save_rng_state.py_impl(FakeTensorMode)
+    def impl_fake_tensor_mode(op, *args, **kwargs):
+        # Check device to call the right impl
+        return impl_backend_select(op, *args, **kwargs)
+
+    @run_and_save_rng_state.py_impl(ProxyTorchDispatchMode)
+    def impl_proxy_dispatch_mode(op, *args, **kwargs):
+        mode = _get_current_dispatch_mode()
+        assert mode is not None
+        with _pop_mode_temporarily() as mode:
+            if mode.enable_tracing:
+                out = impl_fake_tensor_mode(op, *args, **kwargs)
+                proxy_args = pytree.tree_map(partial(unwrap_proxy, mode), (op, *args))
+                proxy_kwargs = pytree.tree_map(partial(unwrap_proxy, mode), kwargs)
+                out_proxy = mode.tracer.create_proxy(
+                    "call_function", run_and_save_rng_state, proxy_args, proxy_kwargs
+                )
+                return track_tensor_tree(
+                    out, out_proxy, constant=None, tracer=mode.tracer
+                )
+            else:
+                return run_and_save_rng_state(op, *args, **kwargs)
+
+
+def register_run_with_rng_state_op():
+    run_with_rng_state = HigherOrderOperator("run_with_rng_state")
+
```

Comment:
How do you access `run_with_rng_state` later? I don't see it being returned from this function

Reply:
Do we have to return? 

`HigherOrderOperator("run_with_rng_state")` seems to register a `torch.ops.run_with_rng_state` already. So, I don't really need to return this.

It could totally be that my understanding is incorrect. So, please correct me if I am wrong.

Reply:
You're right, I didn't realize this is what the code did.